source/common/thread_local/thread_local_impl.cc

```diff
@@ -61,6 +115,38 @@ void InstanceImpl::registerThread(Event::Dispatcher& dispatcher, bool main_threa
   }
 }
 
+// Puts the slot into a deferred delete container, the slot will be destructed when its out-going
+// callback reference count goes to 0.
+void InstanceImpl::recycle(std::unique_ptr<SlotImpl>&& slot) {
+  ASSERT(std::this_thread::get_id() == main_thread_id_);
+  ASSERT(slot != nullptr);
+  auto* slot_addr = slot.get();
+  deferred_deletes_.insert({slot_addr, std::move(slot)});
```

Comment:
I'm not clear on why we need this.  Can't we have both Bookkeeper and threads have pointers to ref_count_, and have scheduleCleanup either straight up delete slot or schedule deletion via post?  Why do we need the intermediate map?

Reply:
when we destruct the Bookkeeper, we can't delete the SlotImpl yet if the ref_count > 1, we need to detach the SlotImpl instance from its owner, and wait for the ref_count to go to 0 and make the post-call.

Reply:
We talked offline, Alyssa suggested to remove the intermediate storage (deferred_deletes_ ) as well, IIUC, by: releasing the SlotImpl from the unique_ptr, having the ref_count shared_ptr post a callback on destruction to the Main dispatcher. There is a risk of leaking at shutdown time though: in Server::InstanceImpl::terminate(), workers are stopped after main dispatcher loop is exited, if a worker posts a cleanup callback, the callback will not be run, the SlotImpl instance pointed to by the raw pointer will be leaked. So the intermediate container is kinda unavoidable.